### PR TITLE
Add support for Map datatype

### PR DESCRIPTION
**Pull request by danielwestendorf**
[Clickhouse Datatype Docs](https://clickhouse.com/docs/en/sql-reference/data-types/map)

Adapted from [this commit](https://github.com/huntresslabs/clickhouse-activerecord/commit/0d9db1d3ab4bac659be4e864e5041bd728807f62) but adapted to be similar to the existing `Array` datatype implementation.

### CODE REVIEW

File: lib/active_record/connection_adapters/clickhouse_adapter.rb [hunk 1]
```diff
@@ -223,6 +228,8 @@ def quote(value)
         case value
         when Array
           '[' + value.map { |v| quote(v) }.join(', ') + ']'
+        when Hash
+          '{' + value.map { |k, v| "#{quote(k)}: #{quote(v)}" }.join(', ') + '}'
```
Comment (PNixx) [inline lib/active_record/connection_adapters/clickhouse_adapter.rb:232]: Maybe `value.to_json` will be better?

Comment (danielwestendorf) [reply lib/active_record/connection_adapters/clickhouse_adapter.rb:232]: That would definitely be cleaner, however, looking at the [rails source](https://github.com/rails/rails/blob/290b8069b98b87717525382ae2bad5cc2fbf3b66/activerecord/lib/active_record/connection_adapters/abstract/quoting.rb#L73) it looks like manual quoting is more about security than anything.

Tracing the postgres adapter's [implementation](https://github.com/ged/ruby-pg/blob/37197594e64f85cc87322ae8ce16096258bd2c2e/lib/pg/text_encoder/json.rb#L10), it looks like they use `JSON.generate`.

```suggestion
          JSON.generate(value, quirks_mode: true)
```

Comment (danielwestendorf) [reply lib/active_record/connection_adapters/clickhouse_adapter.rb:232]: I had to revert this; clickhouse does not always like JSON generated values a67ca2bbec38953f4bf7f5719a58175466c75130

**Comment by danielwestendorf** [comment]
I think this PR needs to bake a bit more. I'm reverting to draft for now.

**Comment by danielwestendorf** [comment]
This has been working well for us.

**Pull request merged**